smartsim/_core/_cli/build.py

```diff
@@ -285,73 +223,236 @@ def build_redis_ai(
+        validate_installed_version=_create_torch_version_validator(
+            with_suffix=device_suffix
+        ),
+        install_on_absent=fetch_if_missing,
+        install_on_conflict=False,
+        verbose=verbose,
+    )
 
-    if not build_env.check_installed("torch", versions.TORCH):
-        inferred_device = infer_torch_device()
-        if (inferred_device == "gpu") and (device == "cpu"):
-            logger.warning("CPU requested, but GPU backend is available")
-        _install_torch_from_pip(versions, device, verbose)
-    # if torch already installed, check the versions to make sure correct
-    # torch version is downloaded for that particular device
-    else:
-        installed = Version_(pkg_resources.get_distribution("torch").version)
-        if device == "gpu":
-            # if torch version is x.x.x+cpu
-            if "cpu" in installed.patch:
-                msg = (
-                    "Torch CPU is currently installed but torch GPU requested. "
-                    "Uninstall all torch packages and run the `smart build` command "
-                    "again to obtain Torch GPU libraries"
-                )
-                logger.warning(msg)
-
-        elif device == "cpu":
-            # if torch version if x.x.x then we need to install the cpu version
-            if "cpu" not in installed.patch and not build_env.is_macos():
-                msg = (
-                    "Torch GPU installed in python environment but requested Torch CPU."
-                    " Run `pip uninstall torch torchvision` and run `smart build` again"
-                )
-                logger.error(msg)  # error because this is usually fatal
-        logger.info(f"Torch {versions.TORCH} installed in Python environment")
+
+def _create_torch_version_validator(
+    with_suffix: str,
+) -> t.Callable[[str, t.Optional[Version_]], bool]:
+    def check_torch_version(package: str, version: t.Optional[Version_]) -> bool:
+        if not BuildEnv.check_installed(package, version):
+            return False
+        # Default check only looks at major/minor version numbers,
+        # Torch requires we look at the patch as well
+        installed = BuildEnv.get_package_version(package)
+        if with_suffix and with_suffix not in installed.patch:
+            # Torch requires that we compare to the patch
+            raise VersionConflictError(
+                package,
+                installed,
+                version or Version_(f"X.X.X{with_suffix}"),
+                msg=(
+                    f"{package}=={installed} does not satisfy device "
+                    f"suffix requirement: {with_suffix}"
+                ),
+            )
+        return True
+
+    return check_torch_version
+
+
+def install_py_onnx_version(
+    fetch_if_missing: bool = False,
+    verbose: bool = False,
+) -> None:
+    """Check Python environment for a compatible ONNX installation"""
+    logger.info("Searching for a compatible ONNX install...")
```

Comment:
What do you think about adding a logging decorator to put on all the methods we do the immediate logging for?

```
@auto_logger("Searching for a compatible ONNX install...", )
def install_py_onnx_version(...):
  if not Versioner.ONNX:
    ....
```

I know that returns are more difficult unless we refactor to reliably raise exceptions, but at least we could start to avoid logger.info @ the start of half the methods..

Reply:
Love it! Little worried about feature creep. Maybe added to the ticket with the "verbose mode log level"?